wolfcrypt/src/port/nxp/se050_port.c

```diff
@@ -378,56 +391,107 @@ void se050_aes_free(Aes* aes)
+    if (aes->ctxInitDone) {
+        sss_symmetric_context_free(&aes->aes_ctx);
 
-    if (status == kStatus_SSS_Success) {
-        status = sss_key_store_allocate(&host_keystore, 55);
+        /* sets back to zero to indicate that a free has been called */
+        aes->ctxInitDone = 0;
     }
 
+    status = sss_key_store_context_init(&host_keystore, cfg_se050_i2c_pi);
+    if (status == kStatus_SSS_Success) {
+        status = sss_key_store_allocate(&host_keystore, SE050_KEYSTOREID_AES);
+    }
     if (status == kStatus_SSS_Success) {
         status = sss_key_object_init(&keyObject, &host_keystore);
     }
-
     if (status == kStatus_SSS_Success) {
         status = sss_key_object_get_handle(&keyObject, aes->keyId);
+        aes->keyId = -1;
     }
+    sss_key_store_erase_key(&host_keystore, &keyObject);
     sss_key_object_free(&keyObject);
 
-    sss_symmetric_context_free(&aes->aes_ctx);
-
     wolfSSL_CryptHwMutexUnLock();
 }
 
 #endif /* !NO_AES */
 
 #ifdef HAVE_ECC
+
+static sss_cipher_type_t se050_map_curve(int curve_id)
+{
+    sss_cipher_type_t curve_type;
+    switch (curve_id) {
+        case ECC_SECP160K1:
+        case ECC_SECP192K1:
+        case ECC_SECP224K1:
+        case ECC_SECP256K1:
+            curve_type = kSSS_CipherType_EC_NIST_K;
+            break;
+        case ECC_BRAINPOOLP160R1:
+        case ECC_BRAINPOOLP192R1:
+        case ECC_BRAINPOOLP224R1:
+        case ECC_BRAINPOOLP256R1:
+        case ECC_BRAINPOOLP320R1:
+        case ECC_BRAINPOOLP384R1:
+        case ECC_BRAINPOOLP512R1:
+            curve_type = kSSS_CipherType_EC_BRAINPOOL;
+            break;
+        case ECC_CURVE_DEF:
+        case ECC_SECP160R1:
+        case ECC_SECP192R1:
+        case ECC_SECP224R1:
+        case ECC_SECP256R1:
+        case ECC_SECP384R1:
+        case ECC_SECP521R1:
+        default:
+            curve_type = kSSS_CipherType_EC_NIST_P;
+            break;
+    }
+    return curve_type;
+}
+
 int se050_ecc_sign_hash_ex(const byte* in, word32 inLen, byte* out,
                          word32 *outLen, struct ecc_key* key)
 {
+    int                 ret = 0;
     sss_status_t        status;
     sss_asymmetric_t    ctx_asymm;
     sss_key_store_t     host_keystore;
     sss_object_t        newKey;
     sss_algorithm_t     algorithm = kAlgorithm_None;
-    int         keyId = se050_allocate_key(SE050_ECC_SIGN);
-    int         keysize = (word32)key->dp->size;
-    int         ret = BAD_MUTEX_E;
+    int                 keySize;
+    int                 keySizeBits;
+
+#ifdef SE050_DEBUG
+    printf("se050_ecc_sign_hash_ex: key %p, in %p (%d), out %p (%d), keyId %d\n",
+        key, in, inLen, out, *outLen, key->keyId);
+#endif
 
     if (cfg_se050_i2c_pi == NULL) {
         return WC_HW_E;
     }
+    if (key->keyId <= 0) {
+        return BAD_FUNC_ARG;
+    }
+
+    keySize = key->dp->size;
+    keySizeBits = keySize * 8;
+    if (keySizeBits > SSS_MAX_ECC_BITS)
```

Comment:
This looks odd.

Reply:
This is my hack for handling the 521 bit curve. I will find a better solution.